### ISSUE TITLE
indent VariableDeclarator and smart-tabs (possible rule change)

### ISSUE THREAD

**Issue by oh-ren** [eslint/eslint#4496]
Derived from https://github.com/eslint/eslint/issues/2077#issuecomment-157352716

When using *smart tabs* (tabs for indent, spaces for alignment) I have to set `indent`'s `VariableDeclarator` to `0`, otherwise, by default, I would get e.g. (first level of indentation): `Expected indentation of 2 tab characters but found 1.` But by setting it to `0`, I'm left without variable declaration indent/alignment validation.

`Indent` should know smart tabs are used, and validate accordingly.

There is the option `smart-tabs` in the `no-mixed-spaces-and-tabs` rule, but that's only to allow spaces after tab indent.

I think the `smart-tabs` option should/can be moved to the 'indent' rule. In fact, I don't think there's much reason to keep `no-mixed-spaces-and-tabs`, as it's *main* purpose could be incorporated in `indent` as well (isn't it?)

One way it could be possible to adapt `indent` for smart-tabs from the user/config side:
* configure `ìndent` as if you're using spaces (so no 'tab' as first option);
* the way `VariableDeclarator` works stays the same;
* add a new optional validation to the 2nd option object: 'smart-tabs'.

So with smart-tabs enabled, by default, `indent` would expect 1 level of variable declaration indent, which would mean the number of spaces as configured.

This might be a bit confusing though, so another solution would be e.g. to make a more severe rule change, making the first option one of `space`, `tab` or `smart-tab`, requiring to define the number of spaces (having meaning for `space` and `smart-tab`) as the value of the 2nd option, or incorporate in option object.

So.. what do you think? :smile:

**Comment by oh-ren** [comment]
Btw; I do have some doubts about  `VariableDeclarator` being connected to indent level / being a multiplier, why not specify exact number of spaces?
When using smart-tabs I always want 4 spaces for `var` and `let` and 6 for `const`. And it gives me trouble when spaces = 4..
BUT I already noted some discussion going on regarding this.. :) #3339

**Comment by nzakas** [comment]
Sorry, I'm having a hard time following. Can you please clarify what it is you are proposing?

**Comment by oh-ren** [comment]
I'm proposing:

1. make the `indent` rule able to validate smart-tabs;
2. make the `no-mixed-spaces-and-tabs` rule obsolete by making it part of the `indent` rule;
3. make `VariableDeclarator` take an absolute value, instead of an indent-width multiplier.

Mainly this issue is about #1 though. #3 is being discussed in issue #3339, but that one is closed already, as the original issue has been resolved. Perhaps a new issue should be created for that.

Currently, when using indentation style 'tab', `VariableDeclarator` validation expects tabs indent. Though when using smart-tabs, once uses *spaces* to do this 'indent' (instead of *indent* we speak of **alignment** in this case (though `SwitchCase` is still indent (as other control structures).. no spaces needed there) ;))

I tried my best to describe the issue as clear as possible (not a native English speaker). Hope you get the idea now after this post.. :wink:

At the bottom of my OP I do some propositions of how the validation of smart-tabs could be configured.

**Comment by nzakas** [comment]
Thanks, that helps. 

`no-mixed-spaces-and-tabs` might be able to be merged into `indent`. I'm not sure changing indent to support smart tabs. @gyandeeps ?

**Comment by gyandeeps** [comment]
**Idea:** Sounds good.

**Implementing:** I have not given it much thought but from the nature of indent calculations and other complicated stuff, seems beyond the reach (Guesstimating here as we have to actually try some implementation before we can say anything).

**Current Status:** We have so many pending items for indent rule (bugs and enhancements), I am not sure if this would ever get done. But if somebody wants to take up this challenge, go for it.

**Biggest TODO for Indent:** Biggest enchacement that is pending right now for indent rule is supporting [multi-line](https://github.com/eslint/eslint/issues/1801).

**Comment by alberto** [comment]
Closed by lack of activity
